core/staking_verifier.go

```diff
@@ -0,0 +1,265 @@
+	if newRate.GT(wrapper.Validator.MaxRate) {
+		return errCommissionRateChangeTooHigh
+	}
+	if chainContext != nil {
+		// TODO: make sure we are reading from the correct snapshot
+		snapshotValidator, err := chainContext.ReadValidatorSnapshot(wrapper.Address)
+		if err != nil {
+			return err
+		}
+		rateAtBeginningOfEpoch := snapshotValidator.Validator.Rate
+
+		if newRate.Sub(rateAtBeginningOfEpoch).Abs().GT(wrapper.Validator.MaxChangeRate) {
+			return errCommissionRateChangeTooFast
+		}
+	}
+	return wrapper.SanityCheck()
+}
+
+// VerifyDelegateStakingMsg verifies the delegate message using the stateDB if
+// it is provided. Otherwise, it will only check components that don't require it.
+//
+// Note that these checks never update the stateDB, it only reads from stateDB.
+func VerifyDelegateStakingMsg(stateDB vm.StateDB, msg *staking.Delegate) error {
+	if msg.Amount.Sign() == -1 {
+		return errNegativeAmount
+	}
+	if stateDB == nil {
+		return nil
+	}
+	if !stateDB.IsValidator(msg.ValidatorAddress) {
+		return errValidatorNotExist
+	}
+	wrapper := stateDB.GetStakingInfo(msg.ValidatorAddress)
+	if wrapper == nil {
+		return errValidatorNotExist
+	}
+	delegatorExist := false
+	for _, delegation := range wrapper.Delegations {
+		if bytes.Equal(delegation.DelegatorAddress.Bytes(), msg.DelegatorAddress.Bytes()) {
+			delegatorExist = true
+			totalInUndelegation := delegation.TotalInUndelegation()
+			balance := stateDB.GetBalance(msg.DelegatorAddress)
+			// If the sum of normal balance and the total amount of tokens in undelegation is greater than the amount to delegate
+			if big.NewInt(0).Add(totalInUndelegation, balance).Cmp(msg.Amount) >= 0 {
+				// First, check if it can use tokens in undelegation to delegate (redelegate)
+				delegateBalance := big.NewInt(0).Set(msg.Amount)
+				// Use the latest undelegated token first as it has the longest remaining locking time.
+				i := len(delegation.Undelegations) - 1
+				for ; i >= 0; i-- {
+					if delegation.Undelegations[i].Amount.Cmp(delegateBalance) <= 0 {
+						delegateBalance.Sub(delegateBalance, delegation.Undelegations[i].Amount)
+					} else {
+						delegation.Undelegations[i].Amount.Sub(delegation.Undelegations[i].Amount, delegateBalance)
+						delegateBalance = big.NewInt(0)
+						break
+					}
+				}
+				delegation.Undelegations = delegation.Undelegations[:i+1]
+				delegation.Amount.Add(delegation.Amount, msg.Amount)
+				if err := wrapper.SanityCheck(); err != nil {
+					return err
+				}
+				// Second, if all locked token are used, check if it can use balance.
+				if delegateBalance.Cmp(big.NewInt(0)) > 0 {
+					return nil
+				}
+				return errInsufficientBalanceForStake // This shouldn't really happen
+			}
+			return errors.Wrapf(
+				errInsufficientBalanceForStake,
+				"total-delegated %s own-current-balance %s amount-to-delegate %s",
+				totalInUndelegation.String(),
+				balance.String(),
+				msg.Amount.String(),
+			)
+		}
+	}
+	if !delegatorExist {
```

Comment:
why not remove delegatorExist here, but in the apply....?

Reply:
oh, this can be removed. Done in latest